src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+                                                     .WithAdditionalAnnotations(Formatter.Annotation);
+                        editor.ReplaceNode(match, replacement);
+                    }
+                }
+                else
+                {
+                    editor.ReplaceNode(match, replacement);
+                }
+            }
+        }
+
+        private int GetInsertionIndex(SemanticDocument document, Compilation compilation, ISyntaxFactsService syntaxFacts, SyntaxNode methodDeclaration, CancellationToken cancellationToken)
+        {
+            var parameterList = syntaxFacts.GetParameterList(methodDeclaration);
+            var parameterSymbolList = GetParameterList(document, parameterList!, cancellationToken);
+            var insertionIndex = 0;
+
+            foreach (var parameterSymbol in parameterSymbolList)
+            {
+                if (!parameterSymbol.HasExplicitDefaultValue && !parameterSymbol.Type.Equals(compilation.GetTypeByMetadataName(typeof(CancellationToken).FullName)))
+                {
+                    insertionIndex++;
+                }
+            }
+
+            return insertionIndex;
+        }
+
+        /// <summary>
+        /// Goes through all of the invocations and replaces the expression with identifiers from the invocation 
+        /// arguments and rewrites the call site with the updated expression as a new argument
+        /// </summary>
+        private async Task<Solution> RewriteCallSitesWithNewMethodAsync(SemanticDocument semanticDocument,
+                                                                               TExpressionSyntax expression,
+                                                                               IMethodSymbol methodSymbol,
+                                                                               bool allOccurrences,
+                                                                               string parameterName,
+                                                                               ImmutableDictionary<Document, List<TInvocationExpressionSyntax>> callSites,
+                                                                               bool trampoline,
+                                                                               CancellationToken cancellationToken)
+        {
+            var firstCallSite = callSites.Keys.First();
+            var currentSolution = firstCallSite.Project.Solution;
+            var modifiedSolution = currentSolution;
+            var newMethodIdentifier = methodSymbol.Name + "_" + parameterName;
+
+            foreach (var grouping in callSites.GroupBy(kvp => kvp.Key.Project))
+            {
+                var project = grouping.Key;
+                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var keyValuePair in grouping)
+                {
+                    var document = keyValuePair.Key;
```

Comment:
can you extract this loop body to it's own helper?

Reply:
extracted